app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/recordaudio/RecordAudioViewModel.kt

```diff
@@ -0,0 +1,319 @@
+import java.io.IOException
+import javax.inject.Inject
+import kotlin.io.path.deleteIfExists
+import kotlin.properties.Delegates
+
+@Suppress("TooManyFunctions")
+interface RecordAudioViewModel {
+    fun getButtonState(): RecordAudioButtonState
+    fun getDiscardDialogState(): RecordAudioDialogState
+    fun getPermissionsDeniedDialogState(): RecordAudioDialogState
+    fun getMaxFileSizeReachedDialogState(): RecordAudioDialogState
+    fun getInfoMessage(): SharedFlow<UIText>
+    fun getOutputFile(): File?
+    fun getAudioState(): AudioState
+    fun startRecording(context: Context)
+    fun stopRecording()
+    fun showDiscardRecordingDialog(onCloseRecordAudio: () -> Unit)
+    fun onDismissDiscardDialog()
+    fun showPermissionsDeniedDialog()
+    fun onDismissPermissionsDeniedDialog()
+    fun onDismissMaxFileSizeReachedDialog()
+    fun discardRecording(onCloseRecordAudio: () -> Unit)
+    fun sendRecording(onAudioRecorded: (UriAsset) -> Unit, onComplete: () -> Unit)
+    fun onPlayAudio()
+    fun onSliderPositionChange(position: Int)
+}
+
+@Suppress("TooManyFunctions")
+@HiltViewModel
+class RecordAudioViewModelImpl @Inject constructor(
+    private val kaliumFileSystem: KaliumFileSystem,
+    private val recordAudioMessagePlayer: RecordAudioMessagePlayer,
+    private val getAssetSizeLimit: GetAssetSizeLimitUseCase,
+    private val observeEstablishedCalls: ObserveEstablishedCallsUseCase,
+    private val currentScreenManager: CurrentScreenManager
+) : RecordAudioViewModel, ViewModel() {
+
+    private var state: RecordAudioState by mutableStateOf(RecordAudioState())
+
+    private var assetLimitInMegabyte by Delegates.notNull<Long>()
+
+    private var mediaRecorder: MediaRecorder? = null
+
+    private var hasOngoingCall: Boolean = false
+
+    private val infoMessage = MutableSharedFlow<UIText>()
+
+    override fun getButtonState(): RecordAudioButtonState = state.buttonState
+
+    override fun getDiscardDialogState(): RecordAudioDialogState = state.discardDialogState
+
+    override fun getPermissionsDeniedDialogState(): RecordAudioDialogState =
+        state.permissionsDeniedDialogState
+
+    override fun getMaxFileSizeReachedDialogState(): RecordAudioDialogState =
+        state.maxFileSizeReachedDialogState
+
+    override fun getOutputFile(): File? = state.outputFile
+
+    override fun getAudioState(): AudioState = state.audioState
+
+    override fun getInfoMessage(): SharedFlow<UIText> = infoMessage.asSharedFlow()
+
+    init {
+        observeAudioPlayerState()
+
+        viewModelScope.launch {
+            assetLimitInMegabyte = getAssetSizeLimit(isImage = false)
+
+            launch {
+                observeAudioFileSize()
+            }
+            launch {
+                observeScreenState()
+            }
+            launch {
+                observeUserIsInCall()
+            }
+        }
+    }
+
+    private suspend fun observeUserIsInCall() {
+        observeEstablishedCalls.invoke().collect {
```

Comment:
why not just `observeEstablishedCalls().collect {` ? (without `invoke`)

Reply:
Guess it just slip by and I did not notice it 😅